When releasing a resize in full screen video, the video content briefly flashes to its original size

rr-codes opened the pull request:
#### 73498188cfbe2ca049303901013363db5e5e43c0
<pre>
When releasing a resize in full screen video, the video content briefly flashes to its original size
<a href="https://bugs.webkit.org/show_bug.cgi?id=260558">https://bugs.webkit.org/show_bug.cgi?id=260558</a>
rdar://113771996

Reviewed by Aditya Keerthi.

When releasing a resize of a window on visionOS that contains a fixed-position element, the element&apos;s size
briefly becomes its original size.

In `_endLiveResize`, we create a snapshot view and put it on top of the web content. The snapshot is then
removed in the completion handler of a `doAfterNextPresentationUpdate` call.

However, by the time the transaction commit happens in the presentation update, two important
updates have yet to happen:
1. A visible content rect update
2. A geometry update

The visible content rect update needs to happen before the commit so that all the new sizes are updated.
However, `doAfterNextPresentationUpdate` does not ensure this. To fix, bundle a VCR update inside the
`UpdateGeometry` message, and then have the web page update the VCRs when receiving the message.

The geometry update of the drawing area was not happening because its size was never being set. When a
geometry uodate is deferred, `_frameOrBoundsMayHaveChanged` skips setting all the relevant new values.
Instead, all these values are (theoretically) set in `_didStopDeferringGeometryUpdates`. However,
`_didStopDeferringGeometryUpdates` was missing setting the size of the drawing area.

As a result, in `sendUpdateGeometry`, the layout size and other changes were being updated, but the
drawing area size was the old value. It then eventually does get updated on the next UIKit layout via
`frameOrBoundsMayHaveChanged`, but by that point the snapshot view has already been removed.

Fix by simply ensuring that `_didStopDeferringGeometryUpdates` also updates the drawing area size.

* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.h:
* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
(-[WKWebView _createVisibleRectUpdateInfo]):
(-[WKWebView _didStopDeferringGeometryUpdates]):
* Source/WebKit/UIProcess/PageClient.h:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.h:
(WebKit::RemoteLayerTreeDrawingAreaProxy::setLastVisibleContentRectUpdate):
(WebKit::RemoteLayerTreeDrawingAreaProxy::lastVisibleContentRectUpdate const):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.mm:
(WebKit::RemoteLayerTreeDrawingAreaProxy::sendUpdateGeometry):
* Source/WebKit/UIProcess/ios/PageClientImplIOS.h:
* Source/WebKit/UIProcess/ios/PageClientImplIOS.mm:
(WebKit::PageClientImpl::createVisibleRectUpdateInfo):
* Source/WebKit/UIProcess/ios/WKContentView.h:
* Source/WebKit/UIProcess/ios/WKContentView.mm:
(-[WKContentView createVisibleRectUpdateInfoFromVisibleRect:unobscuredRect:contentInsets:unobscuredRectInScrollViewCoordinates:obscuredInsets:unobscuredSafeAreaInsets:inputViewBounds:scale:minimumScale:viewStability:enclosedInScrollableAncestorView:sendEvenIfUnchanged:]):
(-[WKContentView didUpdateVisibleRect:unobscuredRect:contentInsets:unobscuredRectInScrollViewCoordinates:obscuredInsets:unobscuredSafeAreaInsets:inputViewBounds:scale:minimumScale:viewStability:enclosedInScrollableAncestorView:sendEvenIfUnchanged:]):
* Source/WebKit/UIProcess/ios/WebPageProxyIOS.mm:
* Source/WebKit/WebProcess/WebPage/DrawingArea.h:
* Source/WebKit/WebProcess/WebPage/DrawingArea.messages.in:
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteLayerTreeDrawingArea.h:
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteLayerTreeDrawingArea.mm:
(WebKit::RemoteLayerTreeDrawingArea::updateGeometry):

Canonical link: <a href="https://commits.webkit.org/267313@main">https://commits.webkit.org/267313@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1e4859e6b5899f689d319edfb8ecc30a4c237b18

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/16245 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/16564 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/16981 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/18011 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/15243 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/19641 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/16677 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/17630 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/16439 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/16886 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/13894 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/18778 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/14129 "Passed tests") | | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/21534 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/15118 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/14872 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/18124 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/15465 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/13111 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/14690 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/3885 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/19056 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/15289 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rr-codes commented on Source/WebKit/UIProcess/API/ios/WKWebViewIOS.h:
```diff
@@ -25,6 +25,7 @@
 
 #import "WKWebViewInternal.h"
 #import "_WKTapHandlingResult.h"
+#import "VisibleContentRectUpdateInfo.h"
```
is this import needed?

rr-codes commented on Source/WebKit/UIProcess/API/ios/WKWebViewIOS.h:
```diff
@@ -52,6 +53,7 @@ enum class TapHandlingResult : uint8_t;
 
 - (void)_frameOrBoundsMayHaveChanged;
 - (BOOL)_shouldDeferGeometryUpdates;
+- (WebKit::VisibleContentRectUpdateInfo)_createVisibleRectUpdateInfo;
```
is the `WebKit::` needed?

rr-codes commented on Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.h:
```diff
@@ -74,7 +75,11 @@ class RemoteLayerTreeDrawingAreaProxy : public DrawingAreaProxy {
     unsigned countOfTransactionsWithNonEmptyLayerChanges() const { return m_countOfTransactionsWithNonEmptyLayerChanges; }
-    
+
+    void setLastVisibleContentRectUpdate(const VisibleContentRectUpdateInfo& info) { m_info = info; }
```
remove

rr-codes commented on Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.h:
```diff
@@ -132,6 +137,8 @@ class RemoteLayerTreeDrawingAreaProxy : public DrawingAreaProxy {
     enum CommitLayerTreeMessageState { CommitLayerTreePending, NeedsDisplayDidRefresh, MissedCommit, Idle };
     CommitLayerTreeMessageState m_commitLayerTreeMessageState { Idle };
 
+    VisibleContentRectUpdateInfo m_info;
```
remove

rr-codes commented on Source/WebKit/WebProcess/WebPage/DrawingArea.h:
```diff
@@ -153,7 +153,7 @@ class DrawingArea : public IPC::MessageReceiver, public WebCore::DisplayRefreshM
 
 #if PLATFORM(COCOA)
-    virtual void updateGeometry(const WebCore::IntSize& viewSize, bool flushSynchronously, const WTF::MachSendRight& fencePort, CompletionHandler<void()>&&) = 0;
+    virtual void updateGeometry(const WebCore::IntSize& viewSize, VisibleContentRectUpdateInfo info, bool flushSynchronously, const WTF::MachSendRight& fencePort, CompletionHandler<void()>&&) = 0;
```
const &

rr-codes commented on Source/WebKit/UIProcess/ios/WKContentView.mm:
```diff
@@ -515,7 +515,7 @@ - (CGRect)_computeUnobscuredContentRectRespectingInputViewBounds:(CGRect)unobscu
 }
 
-- (void)didUpdateVisibleRect:(CGRect)visibleContentRect
+- (WebKit::VisibleContentRectUpdateInfo)createVisibleRectUpdateInfoFromVisibleRect:(CGRect)visibleContentRect
```
is it ok that this isn't a const function?

hortont424 commented on Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
```diff
@@ -2250,6 +2250,18 @@ - (void)_frameOrBoundsWillChange
+            sizeChanged = drawingArea->setSize(WebCore::IntSize(self.bounds.size));
+    }
+
+    if (sizeChanged & [self usesStandardContentView])
```
I think this should be `&&` :) (I realize you're just moving it)

hortont424 commented on Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
```diff
@@ -2755,6 +2781,8 @@ - (void)_didCompleteAnimatedResize
     [self _didStopDeferringGeometryUpdates];
 }
 
+// This should do at least all the things that had been skipped in `-[WKWebView _frameOrBoundsMayHaveChanged]`,
```
If you're going to have this comment, I think it should at least be inside the function. And maybe it'a a reasonable comment? Just unusual.

hortont424 commented on Source/WebKit/UIProcess/PageClient.h:
```diff
@@ -549,6 +550,7 @@ class PageClient : public CanMakeWeakPtr<PageClient> {
     virtual void didFailNavigation(API::Navigation*) = 0;
     virtual void didSameDocumentNavigationForMainFrame(SameDocumentNavigationType) = 0;
 
+
```
?

hortont424 commented on Source/WebKit/UIProcess/PageClient.h:
```diff
@@ -679,6 +681,10 @@ class PageClient : public CanMakeWeakPtr<PageClient> {
 #endif
 
+#if PLATFORM(IOS_FAMILY)
+    virtual std::optional<WebKit::VisibleContentRectUpdateInfo> createVisibleRectUpdateInfo() = 0;
```
Should the function have the full type name in it? Might help with grepping.

smfr commented on Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
```diff
@@ -2530,6 +2535,27 @@ - (BOOL)_shouldDeferGeometryUpdates
+
+    auto contentInsets = [self currentlyVisibleContentInsetsWithScale:scaleFactor obscuredInsets:computedContentInsetUnadjustedForKeyboard];
+
+    return [_contentView createVisibleContentRectUpdateInfoFromVisibleRect:visibleRectInContentCoordinates unobscuredRect:unobscuredRectInContentCoordinates contentInsets:contentInsets unobscuredRectInScrollViewCoordinates:unobscuredRect obscuredInsets:_obscuredInsets unobscuredSafeAreaInsets:[self _computedUnobscuredSafeAreaInset] inputViewBounds:_inputViewBoundsInWindow scale:scaleFactor minimu ...
```
Please put each argument on its own line for our future sanity when we edit this.

rr-codes replied:
I had that originally but then the style checker got mad :( will change it back

smfr commented on Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
```diff
@@ -2759,6 +2785,9 @@ - (void)_didStopDeferringGeometryUpdates
 {
     [self _scheduleVisibleContentRectUpdate];
 
+    // This should do at least all the things that had been skipped in `-[WKWebView _frameOrBoundsMayHaveChanged]`,
```
Can we enforce that in code somehow?

rr-codes replied:
I factored out as much common code as possible, but the rest of it gets intertwined with other properties being set, and I didn't want to change the order of that, so I left as is.

smfr commented on Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.mm:
```diff
@@ -130,7 +131,14 @@
     m_isWaitingForDidUpdateGeometry = true;
-    m_webPageProxy.sendWithAsyncReply(Messages::DrawingArea::UpdateGeometry(m_size, false /* flushSynchronously */, MachSendRight()), [weakThis = WeakPtr { this }] {
+
+#if PLATFORM(IOS_FAMILY)
```
This block would be cleaner as:
```
  std::optional<VisibleContentRectUpdateInfo> visibleRectUpdateInfo;
#if PLATFORM(IOS_FAMILY)
  visibleRectUpdateInfo = m_webPageProxy.pageClient().createVisibleContentRectUpdateInfo();
#endif

smfr commented on Source/WebKit/UIProcess/ios/WKContentView.mm:
```diff
@@ -559,21 +559,42 @@ - (void)didUpdateVisibleRect:(CGRect)visibleContentRect
+    enclosedInScrollableAncestorView:(BOOL)enclosedInScrollableAncestorView
+    sendEvenIfUnchanged:(BOOL)sendEvenIfUnchanged
+{
+    auto visibleContentRectUpdateInfo = [self createVisibleContentRectUpdateInfoFromVisibleRect:visibleContentRect unobscuredRect:unobscuredContentRect contentInsets:contentInsets unobscuredRectInScrollViewCoordinates:unobscuredRectInScrollViewCoordinates obscuredInsets:obscuredInsets unobscuredSafeAreaInsets:unobscuredSafeAreaInsets inputViewBounds:inputViewBounds scale:zoomScale minimumScale:mi ...
```
Maybe the caller can pass a `visibleContentRectUpdateInfo` so we don't repeat the massive list of arguments twice.

rr-codes replied:
unfortunately the passed arguments only add up to a subset of the properties in `WebKit::VisibleContentRectUpdateInfo`, so we can't do this.

pxlcoder commented on Source/WebKit/UIProcess/ios/WKContentView.h:
```diff
@@ -23,6 +23,7 @@
  * THE POSSIBILITY OF SUCH DAMAGE.
  */
 
+#import "VisibleContentRectUpdateInfo.h"
```
Can we forward declare?

pxlcoder commented on Source/WebKit/UIProcess/API/ios/WKWebViewIOS.h:
```diff
@@ -23,6 +23,7 @@
  * THE POSSIBILITY OF SUCH DAMAGE.
  */
 
+#import "VisibleContentRectUpdateInfo.h"
```
Can we forward declare?

rr-codes committed:
When releasing a resize in full screen video, the video content briefly flashes to its original size
https://bugs.webkit.org/show_bug.cgi?id=260558
rdar://113771996

Reviewed by Aditya Keerthi.

When releasing a resize of a window on visionOS that contains a fixed-position element, the element's size
briefly becomes its original size.

In `_endLiveResize`, we create a snapshot view and put it on top of the web content. The snapshot is then
removed in the completion handler of a `doAfterNextPresentationUpdate` call.

However, by the time the transaction commit happens in the presentation update, two important
updates have yet to happen:
1. A visible content rect update
2. A geometry update

The visible content rect update needs to happen before the commit so that all the new sizes are updated.
However, `doAfterNextPresentationUpdate` does not ensure this. To fix, bundle a VCR update inside the
`UpdateGeometry` message, and then have the web page update the VCRs when receiving the message.

The geometry update of the drawing area was not happening because its size was never being set. When a
geometry uodate is deferred, `_frameOrBoundsMayHaveChanged` skips setting all the relevant new values.
Instead, all these values are (theoretically) set in `_didStopDeferringGeometryUpdates`. However,
`_didStopDeferringGeometryUpdates` was missing setting the size of the drawing area.

As a result, in `sendUpdateGeometry`, the layout size and other changes were being updated, but the
drawing area size was the old value. It then eventually does get updated on the next UIKit layout via
`frameOrBoundsMayHaveChanged`, but by that point the snapshot view has already been removed.

Fix by simply ensuring that `_didStopDeferringGeometryUpdates` also updates the drawing area size.

* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.h:
* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
(-[WKWebView _createVisibleRectUpdateInfo]):
(-[WKWebView _didStopDeferringGeometryUpdates]):
* Source/WebKit/UIProcess/PageClient.h:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.h:
(WebKit::RemoteLayerTreeDrawingAreaProxy::setLastVisibleContentRectUpdate):
(WebKit::RemoteLayerTreeDrawingAreaProxy::lastVisibleContentRectUpdate const):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.mm:
(WebKit::RemoteLayerTreeDrawingAreaProxy::sendUpdateGeometry):
* Source/WebKit/UIProcess/ios/PageClientImplIOS.h:
* Source/WebKit/UIProcess/ios/PageClientImplIOS.mm:
(WebKit::PageClientImpl::createVisibleRectUpdateInfo):
* Source/WebKit/UIProcess/ios/WKContentView.h:
* Source/WebKit/UIProcess/ios/WKContentView.mm:
(-[WKContentView createVisibleRectUpdateInfoFromVisibleRect:unobscuredRect:contentInsets:unobscuredRectInScrollViewCoordinates:obscuredInsets:unobscuredSafeAreaInsets:inputViewBounds:scale:minimumScale:viewStability:enclosedInScrollableAncestorView:sendEvenIfUnchanged:]):
(-[WKContentView didUpdateVisibleRect:unobscuredRect:contentInsets:unobscuredRectInScrollViewCoordinates:obscuredInsets:unobscuredSafeAreaInsets:inputViewBounds:scale:minimumScale:viewStability:enclosedInScrollableAncestorView:sendEvenIfUnchanged:]):
* Source/WebKit/UIProcess/ios/WebPageProxyIOS.mm:
* Source/WebKit/WebProcess/WebPage/DrawingArea.h:
* Source/WebKit/WebProcess/WebPage/DrawingArea.messages.in:
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteLayerTreeDrawingArea.h:
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteLayerTreeDrawingArea.mm:
(WebKit::RemoteLayerTreeDrawingArea::updateGeometry):

Canonical link: https://commits.webkit.org/267313@main
